### PR TITLE
[CSM] Add event on_place_node API lua

### PR DESCRIPTION
**Pull request by Dumbeldor**
#5394

### CODE REVIEW

**Comment by Dumbeldor** [comment]
For the moment I return the position of the targeted block. Would it be better to return the position of the block that will be laid?

I wait for your answers before writing the documentation.

**Comment by bigfoot547** [comment]
Could you return a `pointed_thing`?

File: src/script/common/c_content.cpp [hunk 1]
```diff
@@ -1401,3 +1402,41 @@ void read_json_value(lua_State *L, Json::Value &root, int index, u8 recursion)
...
+		push_v3s16(L, pointed.node_undersurface);
+		lua_setfield(L, -2, "under");
+		push_v3s16(L, pointed.node_abovesurface);
+		lua_setfield(L, -2, "above");
+	}
+	else if(pointed.type == POINTEDTHING_OBJECT)
+	{
+		lua_pushstring(L, "object");
+		lua_setfield(L, -2, "type");
+		push_objectrefGet(L, pointed.object_id);
```
Comment (red-001) [inline src/script/common/c_content.cpp]: What happens if a client-side mod tries to use objectref? Does it not cause a segfault?

File: src/script/cpp_api/s_base.cpp [hunk 2]
```diff
@@ -324,16 +325,9 @@ void ScriptApiBase::objectrefGetOrCreate(lua_State *L,
 	}
 }
 
-void ScriptApiBase::objectrefGet(lua_State *L, u16 id)
+void ScriptApiBase::objectrefGet(lua_State *L, const u16 id)
```
Comment (red-001) [inline src/script/cpp_api/s_base.cpp]: Couldn't you just replace calls to `objectrefGet` to  `push_objectrefGet`?

File: src/script/common/c_content.cpp [hunk 3]
```diff
@@ -1392,3 +1402,41 @@ void read_json_value(lua_State *L, Json::Value &root, int index, u8 recursion)
 	}
 	lua_pop(L, 1); // Pop value
 }
+
+void push_pointed_thing(lua_State *L, const PointedThing &pointed)
+{
+	lua_newtable(L);
+	if(pointed.type == POINTEDTHING_NODE)
+	{
+		lua_pushstring(L, "node");
```
Comment (nerzhul) [inline src/script/common/c_content.cpp]: please fix the code style (spaces before if and braces at the end of the condition

File: src/script/cpp_api/s_base.cpp (hunk 2 above)
Comment (nerzhul) [inline src/script/cpp_api/s_base.cpp]: push_objectref is a better name

File: src/script/common/c_content.cpp (hunk 1 above)
Comment (Dumbeldor) [inline src/script/common/c_content.cpp]: Normally no worries.

File: src/script/common/c_content.h [hunk 4]
```diff
@@ -161,6 +164,10 @@ bool               push_json_value           (lua_State *L,
 void               read_json_value           (lua_State *L, Json::Value &root,
                                               int index, u8 recursion = 0);
 
+void               push_pointed_thing        (lua_State *L, const PointedThing &pointed);
+
+void               push_objectRef(lua_State *L, const u16 id);
```
Comment (nerzhul) [inline src/script/common/c_content.h]: The three functions should be static

File: src/script/cpp_api/s_base.cpp [hunk 5]
```diff
@@ -320,20 +321,13 @@ void ScriptApiBase::objectrefGetOrCreate(lua_State *L,
 	if (cobj == NULL || cobj->getId() == 0) {
 		ObjectRef::create(L, cobj);
 	} else {
-		objectrefGet(L, cobj->getId());
+		push_objectref(L, cobj->getId());
 	}
 }
 
-void ScriptApiBase::objectrefGet(lua_State *L, u16 id)
+void ScriptApiBase::push_objectref(lua_State *L, u16 id)
```
Comment (nerzhul) [inline src/script/cpp_api/s_base.cpp]: remove this function and use the underlying call everywhere

File: src/script/common/c_content.cpp [hunk 6]
```diff
@@ -118,6 +119,15 @@ void read_item_definition(lua_State* L, int index,
 }
 
 /******************************************************************************/
+void push_item_definition(lua_State *L, const ItemDefinition &i) {
```
Comment (SmallJoker) [inline src/script/common/c_content.cpp]: Bracket should be on new line

File: src/script/common/c_content.cpp [hunk 7]
```diff
@@ -1427,3 +1437,37 @@ void read_json_value(lua_State *L, Json::Value &root, int index, u8 recursion)
...
+	lua_newtable(L);
+	if (pointed.type == POINTEDTHING_NODE) {
+		lua_pushstring(L, "node");
+		lua_setfield(L, -2, "type");
+		push_v3s16(L, pointed.node_undersurface);
+		lua_setfield(L, -2, "under");
+		push_v3s16(L, pointed.node_abovesurface);
+		lua_setfield(L, -2, "above");
+	}
+	else if (pointed.type == POINTEDTHING_OBJECT) {
```
Comment (SmallJoker) [inline src/script/common/c_content.cpp]: `} else if (...)`

File: src/script/common/c_content.h [hunk 8]
```diff
@@ -162,6 +164,10 @@ bool               push_json_value           (lua_State *L,
...
```
Comment (SmallJoker) [inline src/script/common/c_content.h]: even if this looks (IMO) horrible as it was done before, align it the same way as the other functions, please.

**Comment by sofar** [comment]
Looks all good to me now, @smalljoker?

**Pull request merged**